pkg/cluster/actions.go

```diff
@@ -0,0 +1,231 @@
+// predictable, "kubeadm friendly" and consistent execution order.
+type ExecutionPlan []*PlannedTask
+
+// internal registry of named Action implementations
+var actionImpls = struct {
+	impls map[string]func() Action
+	sync.Mutex
+}{
+	impls: map[string]func() Action{},
+}
+
+// RegisterAction registers a new named actionBuilder function for use
+func RegisterAction(name string, actionBuilderFunc func() Action) {
+	actionImpls.Lock()
+	actionImpls.impls[name] = actionBuilderFunc
+	actionImpls.Unlock()
+}
+
+// GetAction returns one instance of a registered action
+func GetAction(name string) (Action, error) {
+	actionImpls.Lock()
+	actionBuilderFunc, ok := actionImpls.impls[name]
+	actionImpls.Unlock()
+	if !ok {
+		return nil, fmt.Errorf("no Action implementation with name: %s", name)
+	}
+	return actionBuilderFunc(), nil
+}
+
+// NewExecutionPlan creates an execution plan by applying logical step/task
+// defined for each action to the actual cluster topology. As a result task
+// could be executed zero, one or more times according with the target nodes
+// selector defined for each task.
+// The execution plan is ordered, providing a predictable, "kubeadm friendly"
+// and consistent execution order; with this regard please note that the order
+// of actions is important, and it will be respected by planning.
+// TODO(fabrizio pandini): probably it will be necessary to add another criteria
+//     for ordering planned task for the most complex workflows (e.g.
+//     init-join-upgrade and then join again)
+//     e.g. it should be something like "action group" where each action
+//	   group is a list of actions
+func NewExecutionPlan(cfg *config.Config, actionNames []string) (ExecutionPlan, error) {
+	// for each actionName
+	var plan = ExecutionPlan{}
+	for i, name := range actionNames {
+		// get the action implementation instance
+		actionImpl, err := GetAction(name)
+		if err != nil {
+			return nil, err
+		}
+		// for each logical tasks defined for the action
+		for j, t := range actionImpl.Tasks() {
+			// get the list of target nodes in the current topology
+			targetNodes := t.TargetNodes(cfg)
+			for _, n := range targetNodes {
+				// creates the planned task
+				taskContext := &PlannedTask{
+					Node:        n,
+					Task:        t,
+					actionIndex: i,
+					taskIndex:   j,
+				}
+				plan = append(plan, taskContext)
+			}
+		}
+	}
+
+	// sorts the list of planned task ensuring a predictable, "kubeadm friendly"
+	// and consistent execution order
+	sort.Sort(plan)
+	return plan, nil
+}
+
+// Len of the ExecutionPlan.
+// It is required for making ExecutionPlan sortable.
+func (t ExecutionPlan) Len() int {
+	return len(t)
+}
+
+// Less return the lower between two elements of the ExecutionPlan, where the
+// lower element should be executed before the other.
+// It is required for making ExecutionPlan sortable.
+func (t ExecutionPlan) Less(i, j int) bool {
+	return t[i].ExecutionOrder() < t[j].ExecutionOrder()
```

Comment:
Isn't it a bit strange to relay on string sorting here given that `ExecutionOrder()` returns a `string`? Could we have a utility function that takes into account the node `ProvisioningOrder()` and the `actionIndex` and `taskIndex`? https://play.golang.org/p/z8ZGpyl_EPt

Maybe it's minor as long as the order is always preserved.

Reply:
> the order is always preserved

The ordering is 100% deterministic, and this is covered by unit tests as well

Nevertheless, I'm ready to change implementation if this one seems counter-intuitive, but I'm not 100% sure the alternative reads better. see https://gist.github.com/fabriziopandini/e544842b8c38c90974b215167a77d934

Reply:
I'm going to simplify actions quite a bit in a follow-up O(very soon)